File: bin/generate-import-fixtures.py
#!/usr/bin/env python3
"""Erzeugt docs/testdata/import-fixtures/ (Kontakte + Stunden). Spiegel von bin/generate-import-fixtures.php."""
from __future__ import annotations

import csv
import json
import zipfile
from datetime import date, timedelta
from pathlib import Path
from xml.etree.ElementTree import Element, SubElement, ElementTree

ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / "docs" / "testdata" / "import-fixtures"
KONTAKTE = OUT / "kontakte"
STUNDEN = OUT / "stunden"


def roster() -> list[dict]:
    people = [
        ("demo-dup-01", "Anna", "Duplikat", True),
        ("demo-dup-02", "Bernd", "Doppel", True),
        ("demo-ma-03", "Clara", "Fischer", False),
        ("demo-ma-04", "David", "Hoffmann", False),
        ("demo-ma-05", "Elena", "Jung", False),
        ("demo-ma-06", "Felix", "Keller", False),
        ("demo-ma-07", "Greta", "Lange", False),
        ("demo-ma-08", "Hans", "Meier", False),
        ("demo-ma-09", "Ina", "Neumann", False),
        ("demo-ma-10", "Jonas", "Otto", False),
        ("demo-ma-11", "Karla", "Peters", False),
        ("demo-ma-12", "Leon", "Richter", False),
    ]
    out = []
    for i, (login, first, last, dup) in enumerate(people):
        out.append(
            {
                "login": login,
                "first": first,
                "last": last,
                "email": f"{first.lower()}.{last.lower()}@demo-import.ganz-om.invalid",
                "phone": f"030 {1000 + i:04d} {2000 + i:04d}",
                "dup": dup,
            }
        )
    return out


def write_csv(path: Path, headers: list[str], rows: list[list[str]]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", encoding="utf-8-sig", newline="") as fh:
        w = csv.writer(fh, delimiter=";", quoting=csv.QUOTE_MINIMAL)
        w.writerow(headers)
        w.writerows(rows)


def write_json(path: Path, headers: list[str], rows: list[list[str]]) -> None:
    data = {"rows": [{h: (row[i] if i < len(row) else "") for i, h in enumerate(headers)} for row in rows]}
    path.write_text(json.dumps(data, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def write_xml(path: Path, headers: list[str], rows: list[list[str]]) -> None:
    root = Element("records")
    for row in rows:
        el = SubElement(root, "row")
        for i, h in enumerate(headers):
            key = "".join(c if c.isalnum() or c == "_" else "_" for c in h) or "field"
            SubElement(el, key).text = row[i] if i < len(row) else ""
    tree = ElementTree(root)
    tree.write(path, encoding="utf-8", xml_declaration=True)


def col_name(idx: int) -> str:
    n = idx
    s = ""
    while True:
        s = chr(65 + (n % 26)) + s
        n = n // 26 - 1
        if n < 0:
            break
    return s


def write_xlsx(path: Path, headers: list[str], rows: list[list[str]]) -> None:
    all_rows = [headers] + rows
    shared: list[str] = []
    index: dict[str, int] = {}

    def si(v: str) -> int:
        if v not in index:
            index[v] = len(shared)
            shared.append(v)
        return index[v]

    sheet_rows = []
    for r_idx, line in enumerate(all_rows):
        row_num = r_idx + 1
        cells = []
        for c_idx, val in enumerate(line):
            ref = f"{col_name(c_idx)}{row_num}"
            cells.append(f'<c r="{ref}" t="s"><v>{si(str(val))}</v></c>')
        sheet_rows.append(f'<row r="{row_num}">{"".join(cells)}</row>')

    def esc(s: str) -> str:
        return (
            s.replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace('"', "&quot;")
        )

    sst = (
        f'<sst xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" '
        f'count="{len(shared)}" uniqueCount="{len(shared)}">'
        + "".join(f"<si><t>{esc(s)}</t></si>" for s in shared)
        + "</sst>"
    )
    sheet = (
        '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>'
        '<worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main">'
        f'<sheetData>{"".join(sheet_rows)}</sheetData></worksheet>'
    )
    workbook = (
        '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>'
        '<workbook xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" '
        'xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships">'
        '<sheets><sheet name="Import" sheetId="1" r:id="rId1"/></sheets></workbook>'
    )
    rels = (
        '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>'
        '<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">'
        '<Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument" '
        'Target="xl/workbook.xml"/></Relationships>'
    )
    wb_rels = (
        '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>'
        '<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">'
        '<Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/worksheet" '
        'Target="worksheets/sheet1.xml"/>'
        '<Relationship Id="rId2" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/sharedStrings" '
        'Target="sharedStrings.xml"/></Relationships>'
    )
    content_types = (
        '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>'
        '<Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">'
        '<Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml"/>'
        '<Default Extension="xml" ContentType="application/xml"/>'
        '<Override PartName="/xl/workbook.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet.main+xml"/>'
        '<Override PartName="/xl/worksheets/sheet1.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.worksheet+xml"/>'
        '<Override PartName="/xl/sharedStrings.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.sharedStrings+xml"/>'
        "</Types>"
    )

    if path.exists():
        path.unlink()
    with zipfile.ZipFile(path, "w", compression=zipfile.ZIP_DEFLATED) as z:
        z.writestr("[Content_Types].xml", content_types)
        z.writestr("_rels/.rels", rels)
        z.writestr("xl/workbook.xml", workbook)
        z.writestr("xl/_rels/workbook.xml.rels", wb_rels)
        z.writestr("xl/worksheets/sheet1.xml", sheet)
        z.writestr("xl/sharedStrings.xml", sst)


def contact_rows(source: str, people: list[dict]) -> tuple[list[str], list[list[str]]]:
    if source in ("excel", "other"):
        h = ["Anrede", "Vorname", "Nachname", "E-Mail", "Telefon", "Login", "Rolle"]
        r = [["Frau/Herr", p["first"], p["last"], p["email"], p["phone"], p["login"], "mitarbeiter"] for p in people]
        return h, r
    if source == "outlook":
        h = ["Given Name", "Family Name", "E-mail Address", "Business Phone", "Login", "Rolle"]
        r = [[p["first"], p["last"], p["email"], p["phone"], p["login"], "mitarbeiter"] for p in people]
        return h, r
    if source == "google":
        h = ["Given Name", "Family Name", "E-Mail", "Phone 1 - Value", "Login", "Rolle"]
        r = [[p["first"], p["last"], p["email"], p["phone"], p["login"], "mitarbeiter"] for p in people]
        return h, r
    if source == "datev":
        h = ["Vorname", "Nachname", "E-Mail", "Telefon", "Personalnummer", "Rolle"]
        r = [[p["first"], p["last"], p["email"], p["phone"], p["login"], "mitarbeiter"] for p in people]
        return h, r
    if source == "lexware":
        h = ["Vorname", "Nachname", "E-Mail", "Telefon", "Benutzername", "Rolle"]
        r = [[p["first"], p["last"], p["email"], p["phone"], p["login"], "mitarbeiter"] for p in people]
        return h, r
    if source == "sevdesk":
        h = ["Vorname", "Nachname", "email", "Telefon", "Login", "Rolle"]
        r = [[p["first"], p["last"], p["email"], p["phone"], p["login"], "mitarbeiter"] for p in people]
        return h, r
    if source == "shiftbase":
        h = ["Employee", "E-Mail", "Phone", "Personalnummer", "Rolle"]
        r = [[f"{p['first']} {p['last']}", p["email"], p["phone"], p["login"], "mitarbeiter"] for p in people]
        return h, r
    raise ValueError(source)


def weekdays(start: date, end: date) -> list[date]:
    out = []
    d = start
    while d <= end:
        if d.weekday() < 5:
            out.append(d)
        d += timedelta(days=1)
    return out


def main() -> None:
    KONTAKTE.mkdir(parents=True, exist_ok=True)
    STUNDEN.mkdir(parents=True, exist_ok=True)
    people = roster()
    dups = [p for p in people if p["dup"]]
    sources = ["excel", "outlook", "google", "datev", "lexware", "sevdesk", "shiftbase", "other"]

    h, r = contact_rows("excel", dups)
    write_csv(KONTAKTE / "00-seed-duplikate.csv", h, r)

    for src in sources:
        h, r = contact_rows(src, people)
        write_csv(KONTAKTE / f"kontakte-{src}.csv", h, r)

    h, r = contact_rows("excel", people)
    write_csv(KONTAKTE / "kontakte-excel.txt", h, r)
    write_json(KONTAKTE / "kontakte-excel.json", h, r)
    write_xml(KONTAKTE / "kontakte-excel.xml", h, r)
    write_xlsx(KONTAKTE / "kontakte-excel.xlsx", h, r)

    days = weekdays(date(2026, 6, 1), date(2026, 9, 19))
    excel_rows: list[list[str]] = []
    shift_rows: list[list[str]] = []
    crew_rows: list[list[str]] = []
    other_rows: list[list[str]] = []

    for pi, p in enumerate(people):
        for di, day in enumerate(days):
            start = "08:00" if (pi + di) % 2 == 0 else "07:30"
            end = "17:00" if (pi + di) % 3 == 0 else "16:30"
            pause = "30"
            de = day.strftime("%d.%m.%Y")
            iso = day.isoformat()
            name = f"{p['first']} {p['last']}"
            excel_rows.append([p["email"], p["login"], name, de, start, end, pause, ""])
            shift_rows.append([name, p["email"], iso, start, end, pause, ""])
            crew_rows.append([name, p["login"], de, start, end, pause])
            if (pi + di) % 4 == 0:
                other_rows.append([p["login"], de, "", "", "", "8"])
            else:
                other_rows.append([p["login"], de, start, end, pause, ""])

    write_csv(
        STUNDEN / "stunden-excel.csv",
        ["E-Mail", "Login", "Name", "Datum", "Beginn", "Ende", "Pause_Minuten", "Stunden"],
        excel_rows,
    )
    write_xlsx(
        STUNDEN / "stunden-excel.xlsx",
        ["E-Mail", "Login", "Name", "Datum", "Beginn", "Ende", "Pause_Minuten", "Stunden"],
        excel_rows,
    )
    write_csv(
        STUNDEN / "stunden-shiftbase.csv",
        ["Employee", "Email", "Date", "Clocked in", "Clocked out", "Break", "Hours"],
        shift_rows,
    )
    write_csv(
        STUNDEN / "stunden-crewmeister.csv",
        ["Mitarbeiter", "Personalnummer", "Datum", "Kommen", "Gehen", "Pause"],
        crew_rows,
    )
    write_csv(
        STUNDEN / "stunden-other.csv",
        ["Login", "Datum", "Beginn", "Ende", "Pause_Minuten", "Stunden"],
        other_rows,
    )

    print(f"OK: {OUT}")
    print(f"  Roster: {len(people)} (Seed-Duplikate: {len(dups)})")
    print(f"  Werktage: {len(days)} -> Stundenzeilen: {len(days) * len(people)}")


if __name__ == "__main__":
    main()

File: bin/generate-import-fixtures.php
<?php
declare(strict_types=1);

/**
 * Erzeugt Testdateien unter docs/testdata/import-fixtures/ (Kontakte + Stunden).
 * php bin/generate-import-fixtures.php
 */
$root = dirname(__DIR__);
$outBase = $root . '/docs/testdata/import-fixtures';
$kontakteDir = $outBase . '/kontakte';
$stundenDir = $outBase . '/stunden';

foreach ([$kontakteDir, $stundenDir] as $dir) {
    if (!is_dir($dir) && !mkdir($dir, 0755, true) && !is_dir($dir)) {
        fwrite(STDERR, "Verzeichnis nicht anlegbar: {$dir}\n");
        exit(1);
    }
}

/** @return list<array{login: string, first: string, last: string, email: string, phone: string, dup: bool}> */
function fixtureRoster(): array
{
    $people = [
        ['demo-dup-01', 'Anna', 'Duplikat', true],
        ['demo-dup-02', 'Bernd', 'Doppel', true],
        ['demo-ma-03', 'Clara', 'Fischer', false],
        ['demo-ma-04', 'David', 'Hoffmann', false],
        ['demo-ma-05', 'Elena', 'Jung', false],
        ['demo-ma-06', 'Felix', 'Keller', false],
        ['demo-ma-07', 'Greta', 'Lange', false],
        ['demo-ma-08', 'Hans', 'Meier', false],
        ['demo-ma-09', 'Ina', 'Neumann', false],
        ['demo-ma-10', 'Jonas', 'Otto', false],
        ['demo-ma-11', 'Karla', 'Peters', false],
        ['demo-ma-12', 'Leon', 'Richter', false],
    ];
    $out = [];
    foreach ($people as $i => [$login, $first, $last, $dup]) {
        $out[] = [
            'login' => $login,
            'first' => $first,
            'last' => $last,
            'email' => strtolower($first) . '.' . strtolower($last) . '@demo-import.ganz-om.invalid',
            'phone' => sprintf('030 %04d %04d', 1000 + $i, 2000 + $i),
            'dup' => $dup,
        ];
    }

    return $out;
}

/**
 * @param list<string> $headers
 * @param list<list<string>> $rows
 */
function writeCsv(string $path, array $headers, array $rows): void
{
    $fh = fopen($path, 'wb');
    if ($fh === false) {
        throw new RuntimeException('Schreiben fehlgeschlagen: ' . $path);
    }
    fwrite($fh, "\xEF\xBB\xBF");
    fputcsv($fh, $headers, ';', '"', '\\');
    foreach ($rows as $row) {
        fputcsv($fh, $row, ';', '"', '\\');
    }
    fclose($fh);
}

/**
 * @param list<string> $headers
 * @param list<list<string>> $dataRows
 */
function writeJson(string $path, array $headers, array $dataRows): void
{
    $rows = [];
    foreach ($dataRows as $line) {
        $obj = [];
        foreach ($headers as $i => $h) {
            $obj[$h] = $line[$i] ?? '';
        }
        $rows[] = $obj;
    }
    file_put_contents(
        $path,
        json_encode(['rows' => $rows], JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT) . "\n"
    );
}

/**
 * @param list<string> $headers
 * @param list<list<string>> $dataRows
 */
function writeXml(string $path, array $headers, array $dataRows): void
{
    $xml = new SimpleXMLElement('<?xml version="1.0" encoding="UTF-8"?><records/>');
    foreach ($dataRows as $line) {
        $row = $xml->addChild('row');
        foreach ($headers as $i => $h) {
            $key = preg_replace('/[^a-zA-Z0-9_]+/', '_', $h) ?: 'field';
            $row->addChild($key, htmlspecialchars((string) ($line[$i] ?? ''), ENT_XML1 | ENT_QUOTES, 'UTF-8'));
        }
    }
    $dom = new DOMDocument('1.0', 'UTF-8');
    $dom->preserveWhiteSpace = false;
    $dom->formatOutput = true;
    $dom->loadXML($xml->asXML() ?: '<records/>');
    $dom->save($path);
}

/**
 * Minimales OOXML .xlsx (eine Sheet, shared strings).
 *
 * @param list<string> $headers
 * @param list<list<string>> $dataRows
 */
function writeXlsx(string $path, array $headers, array $dataRows): void
{
    if (!class_exists('ZipArchive')) {
        throw new RuntimeException('ZipArchive fehlt — xlsx nicht erzeugbar.');
    }
    $all = array_merge([$headers], $dataRows);
    $shared = [];
    $sharedIndex = [];
    $addShared = static function (string $v) use (&$shared, &$sharedIndex): int {
        if (isset($sharedIndex[$v])) {
            return $sharedIndex[$v];
        }
        $idx = count($shared);
        $shared[] = $v;
        $sharedIndex[$v] = $idx;

        return $idx;
    };

    $sheetRows = '';
    foreach ($all as $rIdx => $line) {
        $rowNum = $rIdx + 1;
        $cells = '';
        foreach ($line as $cIdx => $val) {
            $col = '';
            $n = $cIdx;
            do {
                $col = chr(65 + ($n % 26)) . $col;
                $n = intdiv($n, 26) - 1;
            } while ($n >= 0);
            $si = $addShared((string) $val);
            $ref = $col . $rowNum;
            $cells .= '<c r="' . $ref . '" t="s"><v>' . $si . '</v></c>';
        }
        $sheetRows .= '<row r="' . $rowNum . '">' . $cells . '</row>';
    }

    $sst = '<sst xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" count="'
        . count($shared) . '" uniqueCount="' . count($shared) . '">';
    foreach ($shared as $s) {
        $sst .= '<si><t>' . htmlspecialchars($s, ENT_XML1 | ENT_QUOTES, 'UTF-8') . '</t></si>';
    }
    $sst .= '</sst>';

    $sheet = '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>'
        . '<worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main">'
        . '<sheetData>' . $sheetRows . '</sheetData></worksheet>';

    $workbook = '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>'
        . '<workbook xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"'
        . ' xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships">'
        . '<sheets><sheet name="Import" sheetId="1" r:id="rId1"/></sheets></workbook>';

    $rels = '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>'
        . '<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">'
        . '<Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument"'
        . ' Target="xl/workbook.xml"/></Relationships>';

    $wbRels = '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>'
        . '<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">'
        . '<Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/worksheet"'
        . ' Target="worksheets/sheet1.xml"/>'
        . '<Relationship Id="rId2" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/sharedStrings"'
        . ' Target="sharedStrings.xml"/></Relationships>';

    $contentTypes = '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>'
        . '<Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">'
        . '<Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml"/>'
        . '<Default Extension="xml" ContentType="application/xml"/>'
        . '<Override PartName="/xl/workbook.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet.main+xml"/>'
        . '<Override PartName="/xl/worksheets/sheet1.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.worksheet+xml"/>'
        . '<Override PartName="/xl/sharedStrings.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.sharedStrings+xml"/>'
        . '</Types>';

    if (is_file($path)) {
        unlink($path);
    }
    $zip = new ZipArchive();
    if ($zip->open($path, ZipArchive::CREATE) !== true) {
        throw new RuntimeException('xlsx Zip fehlgeschlagen: ' . $path);
    }
    $zip->addFromString('[Content_Types].xml', $contentTypes);
    $zip->addFromString('_rels/.rels', $rels);
    $zip->addFromString('xl/workbook.xml', $workbook);
    $zip->addFromString('xl/_rels/workbook.xml.rels', $wbRels);
    $zip->addFromString('xl/worksheets/sheet1.xml', $sheet);
    $zip->addFromString('xl/sharedStrings.xml', $sst);
    $zip->close();
}

/**
 * @param list<array{login: string, first: string, last: string, email: string, phone: string, dup: bool}> $roster
 * @return array{0: list<string>, 1: list<list<string>>}
 */
function contactRowsForSource(string $source, array $roster): array
{
    $excelRows = [];
    foreach ($roster as $i => $p) {
        $excelRows[] = [
            ($i % 2 === 0) ? 'Frau' : 'Herr',
            $p['first'],
            $p['last'],
            '',
            $p['email'],
            $p['phone'],
            $p['login'],
            '',
            'mitarbeiter',
        ];
    }

    return match ($source) {
        'excel', 'other' => [
            ['Anrede', 'Vorname', 'Nachname', 'Firma', 'E-Mail', 'Telefon', 'Login', 'Kundennummer', 'Rolle'],
            $excelRows,
        ],
        'outlook' => [
            ['Given Name', 'Family Name', 'Company', 'E-mail Address', 'Business Phone', 'Login', 'Kundennummer', 'Rolle'],
            array_map(static fn (array $p): array => [
                $p['first'], $p['last'], '', $p['email'], $p['phone'], $p['login'], '', 'mitarbeiter',
            ], $roster),
        ],
        'google' => [
            ['Given Name', 'Family Name', 'Organization Name', 'E-Mail', 'Phone 1 - Value', 'Login', 'Kundennummer', 'Rolle'],
            array_map(static fn (array $p): array => [
                $p['first'], $p['last'], '', $p['email'], $p['phone'], $p['login'], '', 'mitarbeiter',
            ], $roster),
        ],
        'datev' => [
            ['Vorname', 'Nachname', 'Firma', 'E-Mail', 'Telefon', 'Login', 'Mitarbeiternummer', 'Rolle'],
            array_map(static fn (array $p): array => [
                $p['first'], $p['last'], '', $p['email'], $p['phone'], $p['login'], '', 'mitarbeiter',
            ], $roster),
        ],
        'lexware' => [
            ['Vorname', 'Nachname', 'Firma', 'E-Mail', 'Telefon', 'Benutzername', 'Kundennummer', 'Rolle'],
            array_map(static fn (array $p): array => [
                $p['first'], $p['last'], '', $p['email'], $p['phone'], $p['login'], '', 'mitarbeiter',
            ], $roster),
        ],
        'sevdesk' => [
            ['Vorname', 'Nachname', 'company', 'email', 'Telefon', 'Login', 'Kundennummer', 'Rolle'],
            array_map(static fn (array $p): array => [
                $p['first'], $p['last'], '', $p['email'], $p['phone'], $p['login'], '', 'mitarbeiter',
            ], $roster),
        ],
        'shiftbase' => [
            ['Employee', 'Company', 'E-Mail', 'Phone', 'Personalnummer', 'Kundennummer', 'Rolle'],
            array_map(static fn (array $p): array => [
                $p['first'] . ' ' . $p['last'], '', $p['email'], $p['phone'], $p['login'], '', 'mitarbeiter',
            ], $roster),
        ],
        default => throw new InvalidArgumentException('Unbekannte Quelle: ' . $source),
    };
}

/**
 * @return list<string> Y-m-d weekdays inclusive
 */
function weekdaysBetween(string $from, string $to): array
{
    $out = [];
    $d = new DateTimeImmutable($from);
    $end = new DateTimeImmutable($to);
    while ($d <= $end) {
        $n = (int) $d->format('N');
        if ($n <= 5) {
            $out[] = $d->format('Y-m-d');
        }
        $d = $d->modify('+1 day');
    }

    return $out;
}

$roster = fixtureRoster();
$dups = array_values(array_filter($roster, static fn (array $p): bool => $p['dup']));
$sources = ['excel', 'outlook', 'google', 'datev', 'lexware', 'sevdesk', 'shiftbase', 'other'];

// Seed
[$seedH, $seedR] = contactRowsForSource('excel', $dups);
writeCsv($kontakteDir . '/00-seed-duplikate.csv', $seedH, $seedR);

foreach ($sources as $src) {
    [$h, $r] = contactRowsForSource($src, $roster);
    writeCsv($kontakteDir . '/kontakte-' . $src . '.csv', $h, $r);
}

// Format-Pack excel
[$exH, $exR] = contactRowsForSource('excel', $roster);
writeCsv($kontakteDir . '/kontakte-excel.txt', $exH, $exR);
writeJson($kontakteDir . '/kontakte-excel.json', $exH, $exR);
writeXml($kontakteDir . '/kontakte-excel.xml', $exH, $exR);
writeXlsx($kontakteDir . '/kontakte-excel.xlsx', $exH, $exR);

// Stunden
$days = weekdaysBetween('2026-06-01', '2026-09-19');
$excelHoursRows = [];
$shiftbaseRows = [];
$crewRows = [];
$otherRows = [];

foreach ($roster as $pi => $p) {
    foreach ($days as $di => $day) {
        // leichte Variation: Start 08:00 oder 07:30, Ende 16:30 / 17:00
        $start = ($pi + $di) % 2 === 0 ? '08:00' : '07:30';
        $end = ($pi + $di) % 3 === 0 ? '17:00' : '16:30';
        $pause = '30';
        $de = DateTimeImmutable::createFromFormat('Y-m-d', $day)?->format('d.m.Y') ?? $day;

        $excelHoursRows[] = [$p['email'], $p['login'], $p['first'] . ' ' . $p['last'], $de, $start, $end, $pause, ''];
        $shiftbaseRows[] = [
            $p['first'] . ' ' . $p['last'],
            $p['email'],
            $day,
            $start,
            $end,
            $pause,
            '',
        ];
        $crewRows[] = [
            $p['first'] . ' ' . $p['last'],
            $p['login'],
            $de,
            $start,
            $end,
            $pause,
        ];
        // other: abwechselnd Zeiten vs. nur Stunden
        if (($pi + $di) % 4 === 0) {
            $otherRows[] = [$p['login'], $de, '', '', '', '8'];
        } else {
            $otherRows[] = [$p['login'], $de, $start, $end, $pause, ''];
        }
    }
}

writeCsv(
    $stundenDir . '/stunden-excel.csv',
    ['E-Mail', 'Login', 'Name', 'Datum', 'Beginn', 'Ende', 'Pause_Minuten', 'Stunden'],
    $excelHoursRows
);
writeXlsx(
    $stundenDir . '/stunden-excel.xlsx',
    ['E-Mail', 'Login', 'Name', 'Datum', 'Beginn', 'Ende', 'Pause_Minuten', 'Stunden'],
    $excelHoursRows
);
writeCsv(
    $stundenDir . '/stunden-shiftbase.csv',
    ['Employee', 'Email', 'Date', 'Clocked in', 'Clocked out', 'Break', 'Hours'],
    $shiftbaseRows
);
writeCsv(
    $stundenDir . '/stunden-crewmeister.csv',
    ['Mitarbeiter', 'Personalnummer', 'Datum', 'Kommen', 'Gehen', 'Pause'],
    $crewRows
);
writeCsv(
    $stundenDir . '/stunden-other.csv',
    ['Login', 'Datum', 'Beginn', 'Ende', 'Pause_Minuten', 'Stunden'],
    $otherRows
);

echo "OK: Kontakte + Stunden unter docs/testdata/import-fixtures/\n";
echo '  Roster: ' . count($roster) . " (davon Duplikat-Seed: " . count($dups) . ")\n";
echo '  Werktage: ' . count($days) . ' -> Stundenzeilen ca. ' . (count($days) * count($roster)) . "\n";
